rllib/connectors/connector_v2.py

```diff
@@ -149,8 +127,232 @@ def __call__(
+        zip_with_batch_column: Optional[Union[List[Any], Dict[tuple, Any]]] = None,
+    ) -> Iterator[SingleAgentEpisode]:
+        """An iterator over a list of episodes yielding always SingleAgentEpisodes.
+
+        In case items in the list are MultiAgentEpisodes, these are broken down
+        into their individual agents' SingleAgentEpisodes and those are then yielded
+        one after the other.
+
+        Useful for connectors that operator on both single-agent and multi-agent
+        episodes.
+
+        Args:
+            episodes: The list of SingleAgent- or MultiAgentEpisode objects.
+            zip_with_batch_column: If provided, must be a list of batch items
+                corresponding to the given `episodes` (single agent case) or a dict
+                mapping (agentID, moduleID) tuples to lists of individual batch items
+                corresponding to this agent/module combination. The iterator will then
+                yield tuples of SingleAgentEpisode objects (1st item) along with the
+                data item (2nd item) that this episode was responsible for generating
+                originally.
+
+        Yields:
+            All SingleAgentEpisodes in the input list, whereby MultiAgentEpisodes will
+            be broken down into their individual SingleAgentEpisode components.
+        """
+        # Single-agent case.
+        if isinstance(episodes[0], SingleAgentEpisode):
+            if zip_with_batch_column is not None:
+                if len(zip_with_batch_column) != len(episodes):
+                    raise ValueError(
+                        "Invalid `zip_with_batch_column` data: Must have the same "
+                        f"length as the list of episodes ({len(episodes)}), but has "
+                        f"length {len(zip_with_batch_column)}!"
+                    )
+                for episode, data in zip(episodes, zip_with_batch_column):
+                    yield episode, data
+            else:
+                for episode in episodes:
+                    yield episode
+            return
+
+        # Multi-agent case.
+        list_indices = defaultdict(int)
+        for episode in episodes:
+            for sa_episode in episodes.agent_episodes.values():
+                if zip_with_batch_column:
+                    key = (sa_episode.agent_id, sa_episode.module_id)
+                    if len(data[key]) <= list_indices[key]:
+                        raise ValueError(
+                            "Invalid `zip_with_batch_column` data: Must structurally "
+                            "match the single-agent contents in the given list of "
+                            "(multi-agent) episodes!"
+                        )
+                    d = data[key][list_indices[key]]
+                    list_indices[key] += 1
+                    yield sa_episode, d
+                else:
+                    yield sa_episode
+
+    @staticmethod
+    def add_batch_item(
```

Comment:
Convenience API for adding data items from a single_agent episode (maybe from within a multi-agent one) to the batch.